File: GoalSetting.cpp
#include "GoalSetting.hpp"
#include <stdexcept>
#include <iostream>

GoalSetting::GoalSetting(const std::string& id, const std::string& userId, GoalType goalType, double targetValue, const std::string& startDate, const std::string& deadline)
    : id(id), userId(userId), goalType(goalType), targetValue(targetValue), startDate(startDate), deadline(deadline), initialValue(0.0) {}

GoalSetting::GoalSetting(const GoalSetting& other)
    : id(other.id), userId(other.userId), goalType(other.goalType), targetValue(other.targetValue), startDate(other.startDate), deadline(other.deadline), initialValue(other.initialValue) {}

double GoalSetting::updateProgress(double currentValue) {
    try {
        if (this->initialValue == 0.0) return 0.0;
        if (this->initialValue == this->targetValue) throw std::invalid_argument("Начальное и целевое равны");
        double progress = 0.0;
        if (this->goalType == GoalType::WeightLoss) {
            progress = (this->initialValue - currentValue) / (this->initialValue - this->targetValue) * 100.0;
        }
        return progress;
    }
    catch (const std::invalid_argument& ex) {
        std::cerr << "Ошибка прогресса: " << ex.what() << std::endl;
        return 0.0;
    }
}

bool GoalSetting::isAchieved() { return false; }

void GoalSetting::extendDeadline(const std::string& newDate) {
    this->deadline = newDate + " (продлён)"; 
    size_t pos = this->deadline.find("продлён"); 
    if (pos != std::string::npos) std::cout << "Найдено: " << this->deadline.substr(pos) << std::endl; 
}

void GoalSetting::setInitialValue(double val) { this->initialValue = val; }

GoalSetting GoalSetting::operator+(const GoalSetting& other) const {
    GoalSetting combined = *this; 
    combined.targetValue += other.targetValue;
    return combined;
}

double GoalSetting::operator[](int index) const {
    if (index == 0) return this->targetValue;
    if (index == 1) return this->initialValue;
    throw std::out_of_range("Индекс вне диапазона");
}

File: FitnessAppCsharp/GoalSetting.cs
using System;

namespace FitnessApp
{
    public class GoalSetting
    {
        private string id;
        private string userId;
        private GoalType goalType;
        private double targetValue; 
        private string startDate;
        private string deadline;
        private double initialValue;

        public GoalSetting() { }

        public GoalSetting(string id, string userId, GoalType goalType, double targetValue, string startDate, string deadline)
        {
            this.id = id;
            this.userId = userId;
            this.goalType = goalType;
            this.TargetValue = targetValue; 
            this.startDate = startDate;
            this.deadline = deadline;
            this.initialValue = 0.0;
        }

        public double TargetValue
        {
            get { return targetValue; }
            set
            {
                if (value <= 0) throw new ArgumentOutOfRangeException("TargetValue", "Целевое значение должно быть положительным");
                targetValue = value;
            }
        }

        public double UpdateProgress(double currentValue)
        {
            try
            {
                if (initialValue == 0.0) return 0.0;
                if (initialValue == targetValue) throw new InvalidOperationException("Начальное и целевое значения не могут быть равны - деление на ноль");
                double progress = 0.0;
                if (goalType == GoalType.WeightLoss)
                {
                    progress = (initialValue - currentValue) / (initialValue - targetValue) * 100.0;
                }
                return progress;
            }
            catch (InvalidOperationException ex)
            {
                Console.WriteLine($"Ошибка обновления прогресса: {ex.Message}");
                return 0.0;
            }
        }

        public bool IsAchieved() { return false; }

        public void ExtendDeadline(string newDate) { deadline = newDate; }

        public void SetInitialValue(double val) { initialValue = val; }
    }
}